File: Lasmart.Web/wwwroot/ts/services/BaseService.ts
import {IBaseService} from "./IServices/IBaseService";
import {ApiRequest} from "../Models/ApiRequest";
import {ResponseDto} from "../Models/ResponseDto";

export class BaseService implements IBaseService {

    readonly baseUrl: string;
    constructor(baseUrl: string) {
        this.baseUrl = baseUrl;
    }
    sendAsync(apiRequest: ApiRequest): Promise<ResponseDto> {
        return fetch(this.baseUrl + apiRequest.urn, {
            body: JSON.stringify(apiRequest.data),
            method: apiRequest.apiMethod.toString(),
            headers: { "Accept": "application/json", "Content-Type": "application/json" }
        })
            .then(x => x.json())
            .then(x => Object.assign(Array<ResponseDto>, x));
    }

}

File: Lasmart.Web/Services/BaseService.cs
using System.Text;
using Lasmart.Web.Models;
using Lasmart.Web.Services.IServices;
using Newtonsoft.Json;

namespace Lasmart.Web.Services;

public class BaseService : IBaseService
{
    private readonly IHttpClientFactory _clientFactory;

    public BaseService(IHttpClientFactory clientFactory)
    {
        _clientFactory = clientFactory;
    }

    public async Task<ResponseDto> SendAsync(ApiRequest apiRequest)
    {
        try
        {
            var client = _clientFactory.CreateClient("LasmartPointAPI");
            client.DefaultRequestHeaders.Clear();

            var message = new HttpRequestMessage();
            message.Headers.Add("Accept", "application/json");
            message.RequestUri = new Uri(apiRequest.Url);
            if (apiRequest.Data != null)
            {
                message.Content = new StringContent(
                    JsonConvert.SerializeObject(apiRequest.Data),
                    Encoding.UTF8,
                    "application/json");
            }

            switch (apiRequest.ApiMethod)
            {
                case SD.ApiMethod.DELETE:
                    message.Method = HttpMethod.Delete;
                    break;
                default:
                    message.Method = HttpMethod.Get;
                    break;
            }

            var apiResponse = await client.SendAsync(message);

            var apiContent = await apiResponse.Content.ReadAsStringAsync();
            var apiResponseDto = JsonConvert.DeserializeObject<ResponseDto>(apiContent);
            if (apiResponseDto is null)
                throw new NullReferenceException("API response is null!");
            return apiResponseDto;
        }
        catch (Exception ex)
        {
            var apiResponseDto = new ResponseDto
            {
                Message = "Service Error!",
                ErrorMessages = new List<string> { ex.Message },
                IsSuccess = false
            };
            return apiResponseDto;
        }
    }
}